Add option with_short_file_names

petm-dec opened the pull request:
Hello, we use very long paths.
Our other glog libraries (go and c++) log only the file name and not the full path.
I'd like to have the same for our rust logs, so I made this small addition.

davidbarsky commented on src/format.rs:
```diff
@@ -208,13 +208,26 @@ pub(crate) struct FormatProcessData<'a> {
+                        .unwrap_or_default()
+                } else {
+                    f
+                }
```
I'm not opposed to this change, but I'd prefer an approach like this instead:

```suggestion
            .map(|f| {
                if self.with_short_file_names {
                    let f = Path::new(f);
                    f.file_name().map(OsStr::to_str);
                } else {
                    f
                }
```

I haven't compiled this, so it might fail, but I'd rather be operating in terms of `Path`s instead of string manipulation.

davidbarsky commented on src/lib.rs:
```diff
@@ -185,6 +187,13 @@ impl<T> Glog<T> {
         }
     }
 
+    pub fn with_short_file_names(self, with_short_file_names: bool) -> Glog<T> {
```
I understand that this behavior is meant to match `glog`'s, but I can see "short" being somewhat ambiguous (e.g., what qualifies as "short"? A single directory preceding it?).

How would you feel about renaming this to `with_trimmed_directory` and maybe adding a seperate builder option that exposes [`Path::strip_prefix`](https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.strip_prefix), allowing for more fine-grained rendering of paths?

petm-dec replied:
You're right, I'll use Path and rename.

Not sure about using `Path::strip_prefix`. Could be useful, though with the module shown (`with_target(true)`) we would never use it. Either the full path or just the filename, I doubt that stripping away just a part makes much sense.

jsgf replied:
For Rust there's a very high likelihood that the source will be called `lib.rs` (or maybe `mod.rs`) and even with one directory element you'll get `src/lib.rs`. Are you sure that this change will actually give you useful logs?

petm-dec replied:
In combination with `with_target(true)`, otherwise not.

petm-dec commented:
I renamed the option and used Path to strip the directory.
I think an option to only strip a prefix could be added later, not sure if that would be useful for anyone (not to me I think).
Thanks for looking at this.

jsgf commented on src/format.rs:
```diff
@@ -208,13 +208,28 @@ pub(crate) struct FormatProcessData<'a> {
+                    let path = Path::new(f);
+                    path.file_name()
+                        .map(OsStr::to_str)
+                        .unwrap_or_default()
+                        .unwrap_or_default()
```
Wouldn't it be better to use the original full filename if something goes wrong here?

petm-dec replied:
Yes that is better. I added a fixup that does this.

petm-dec commented:
I also added `with_strip_prefix` option as suggested. 

I added a commit on a test branch, with a little example that shows the resulting logs of various cases:
https://github.com/petm-dec/tracing-glog/commit/adc58353a8f4a4f5676e06075265f0c8ac824b21

davidbarsky commented:
Sorry, I should've mentioned this last week when I first saw this PR, but I only thought of this approach last night. How would you feel about a trait called `FormatFile` (inspired by tracing-subscriber's [`FormatTime`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/time/trait.FormatTime.html), with the `self` corresponding to the struct is required to contain the file name, line number, and module path. You'd then be able to define a formatter for how files/targets/modules are rendered without needing to figuring how the setting should be named.

If you're interested, I can make a PR with the above?

petm-dec commented:
> If you're interested, I can make a PR with the above?

Sure, you modify this as you wish. 
FormatFile seems like a nice and flexible approach.
